src/backend/cuda/kernel/memcopy.hpp

```diff
@@ -17,86 +17,248 @@
+            }
+            --mndims;
+            mdims[mndims] = 1;
+            if (RESHAPE) odims[mndims] = 1;
+            --c;  // Redo this colum, since it is removed now
+        }
+    }
+}
 
-template<typename T>
-void memcopy(Param<T> out, CParam<T> in, const dim_t ndims) {
-    auto memCopy = common::getKernel("cuda::memcopy", {memcopy_cuh_src},
-                                     {TemplateTypename<T>()});
+// To increase the workload inside a kernel, we move a part of the ndims
+// dimension to the last one.  Since this not covered by WG or warp, this is
+// always executed as an internal loop.
+// In case of copy operation, input & output parameters needs to be provided!!
+// main dims: mdims, mstrides, mndims
+// optional dims: odims, ostrides
+// ALL parameters will be updated!!
+void inline increaseWorkload(dim_t elements, dim_t mdims[4], dim_t mstrides[4],
+                             dim_t &mndims, dim_t ostrides[4]) noexcept {
+    if (elements >= 8192 * 2 && mndims != AF_MAX_DIMS && mndims != 0) {
+        // Start only increasing the workload, when all available threads are
+        // occupied.
 
-    dim3 threads(DIMX, DIMY);
+        // list is sorted according to performance improvement
+        // 3x looping is faster than 4x, 2x remains faster than no looping
+        for (const dim_t i : {3, 4, 5, 7, 11, 2}) {
+            if (elements >= 8192 * i && (mdims[mndims - 1] % i) == 0) {
+                mdims[mndims - 1] /= i;
+                mdims[AF_MAX_DIMS - 1] = i;
+                const dim_t mstride = mdims[mndims - 1] * mstrides[mndims - 1];
+                const dim_t ostride = mdims[mndims - 1] * ostrides[mndims - 1];
+                for (dim_t c = mndims; c < AF_MAX_DIMS; ++c) {
+                    mstrides[c] = mstride;
+                    ostrides[c] = ostride;
+                }
+                mndims = AF_MAX_DIMS;
+                break;  // Perform this operation only once.
+            }
+        }
+    }
+}
 
-    if (ndims == 1) {
-        threads.x *= threads.y;
-        threads.y = 1;
+template<typename T>
+void memcopy(Param<T> out, CParam<T> in, dim_t indims) {
+    bool isLinear  = true;
+    dim_t elements = (indims == 0) ? 0 : 1;
+    for (dim_t dim = 0; dim < indims; ++dim) {
+        isLinear &=
+            (elements == in.strides[dim]) & (elements == out.strides[dim]);
+        elements *= in.dims[dim];
     }
+    if (elements > 0) {
+        const auto stream = cuda::getActiveStream();
+        if (isLinear) {
+            CUDA_CHECK(cudaMemcpyAsync(out.ptr, in.ptr,
+                                       in.elements() * sizeof(T),
+                                       cudaMemcpyDeviceToDevice, stream));
+        } else {
+            const int *maxGridSize =
+                cuda::getDeviceProp(cuda::getActiveDeviceId()).maxGridSize;
+            serializeArray(in.dims, in.strides, indims, out.strides);
+            increaseWorkload(elements, in.dims, in.strides, indims,
```

Comment:
The linearizeShape is a template so moving the definition, will have no impact on final code length.
Although the CUDA and OPENCL have the same name, there are small differences (types).
Moving all to dim4 would be conversion for all (CUDA uses dim_t, OPENCL uses int[4]).
Versions in jit.cpp are also different from memcopy.hpp

For OPENCL I switched to int, resulting from the performance test results.
As last, I'm somewhat afraid to give these functions a general visibility since they are so extremely specialized, resulting in those small differences.